Source/src/components/ComponentTransform.h

```diff
@@ -7,80 +7,76 @@ namespace Hachiko
-        }
-
-        void SetPosition(float3 new_position);
-        void SetRotationAxis(float3 x, float3 y, float3 z);
-        void LookAt(float3 target, float3 world_up = float3::unitY);
-
-        void SetLocalTransform(float4x4 new_transform);
-        void SetLocalTransform(float3 position, Quat rotation, float3 scale);
-
-        inline void SetLocalPosition(float3 new_position);
-        inline void SetLocalScale(float3 new_scale);
-        inline void SetLocalRotation(Quat new_rotation);
-        inline void SetLocalRotation(float3 rotation_angles);
-
-        void SetGlobalTransform(float4x4 transform);
-        void UpdateGlobalTransformHierarchy();
-
-        void OnTransformUpdated() override
-        {
-            changed = false;
-        }
-
-        [[nodiscard]] bool HasChanged() const
-        {
-            return changed;
-        }
-
-        [[nodiscard]] float4x4 GetTransform() const
+        enum class MatrixCalculationMode
         {
-            return transform;
-        }
-
-        [[nodiscard]] float3 GetPosition() const
-        {
-            return transform.TranslatePart();
-        }
-
-        [[nodiscard]] float3 GetFwd() const
-        {
-            return transform.WorldZ();
-        }
-
-        [[nodiscard]] float3 GetUp() const
-        {
-            return transform.WorldY();
-        }
-
-        [[nodiscard]] float3 GetRight() const
-        {
-            return transform.WorldX();
-        }
+            LOCAL_FROM_GLOBAL,
+            GLOBAL_FROM_LOCAL
+        };
 
+    public:
+        ComponentTransform(GameObject* container);
+        ComponentTransform(GameObject* new_object, const float3& new_position_local, const Quat& new_rotation_local, const float3& new_scale_local);
+        ComponentTransform(GameObject* new_object, const float4x4& matrix);
+        ~ComponentTransform() override = default;
+
+        void OnTransformUpdated() override;
+        
+        static Quat SimulateLookAt(const math::float3& direction);
+        void LookAtTarget(const float3& target);
+        void LookAtDirection(const float3& direction);
+
+        void SetGlobalTransform(const float4x4& new_matrix);
+        void SetPosition(const float3& new_position);
+        void SetScale(const float3& new_scale);
+        void SetRotationAxis(const float3& new_right, const float3& new_up, const float3& new_front);
+        void SetRotation(const Quat& new_rotation);
+        void SetRotationInEulerAngles(const float3& new_rotation_euler);
+        void SetLocalTransform(const float4x4& new_transform_local);
+        void SetLocalTransform(const float3& new_position_local, const Quat& new_rotation_local, const float3& new_scale_local);
+        void SetLocalPosition(const float3& new_position_local);
+        void SetLocalScale(const float3& new_scale_local);
+        void SetLocalRotation(const Quat& new_rotation_local);
+        void SetLocalRotationInEulerAngles(const float3& new_rotation_euler_local);
```

Comment:
Consider to inline some small functions such as

- void LookAtDirection(const float3& direction);
- void SetLocalTransform(const float3& new_position_local, const Quat& new_rotation_local, const float3& new_scale_local);
- Pretty much all the getters
- Any other 1-2 lines, simple functions that I may have missed.

Reply:
Actually, we can talk about this, I'm kinda hesitant to inline any function, as people say compiler already inlines functions it sees as an optimization, and un-inlines any function that it sees as unnecessary. So I didn't know what to do to be honest and went with a header that looks clean. Let's have a chat about this on discord and I will do the necessary changes we discussed.

Reply:
Sure, I'm down to talk anytime. I didn't know about compiler automatically making some functions inline but I will inform myself about how exactly this occurs.

Worst case scenario we are doing something the compiler will end up doing itself. I asked Carlos and Marc a few weeks ago about how often we should be declaring functions as inline, I was worried the executable file would grow too big if I abused it for all the getters/setters. What they told me is that, in our case since we are making a small game for computers we should not have any problem as that is something you usually have to be wary of when making games for limited platforms such as consoles or mobile.

What we can do though is use the profiler to see if by declaring as inline some of the most used getters/setters it yields any positive results.

Reply:
For now, I'm converting all the getters into inline if that's okay? Afterwards we can discuss on which methods to inline as well, but probably compiler will do the inlining/un-inlining to them itself.

Reply:
Perfect!